neps/nep-0364.md

```diff
@@ -0,0 +1,180 @@
+
+Adding existing proved and vetted crypto crates into the runtime is a safe workaround. It will boost performance
+between 20-25x according to our benchmarks. This will both reduce operating costs significantly and will also
+enable the contract to verify all the signatures in one transaction, which will simplify the contract design.
+
+### What other designs have been considered and what is the rationale for not choosing them?
+
+One possible alternative would be to improve the runtime implementation so that it can compile WASM code to a sufficiently
+fast machine code. Even when it may not be as fast as LLVM native produced code it could still be acceptable for
+these types of use cases (CPU intensive functions) and will certainly avoid the need of adding host functions.
+The effort of adding such a feature will be significantly higher than adding these host functions one by one.
+But on the other side, it will decrease the need of including more host functions in the future.
+
+Another alternative is to deal with the high cost of computing/verifying these signatures in some other manner.
+Decreasing the overall cost of gas and increasing the limits of gas available to attach to the contract could be a possibility.
+Introducing such modification for some contracts, and not for some others can be rather arbitrary
+and not straightforward in the implementation, but an alternative nevertheless.
+
+## Specification
+
+This NEP aims to introduce the following host function:
+
+```rust
+/// Ed25519 is a public-key signature system with several attractive features
+///
+/// Proof of Stake Validator sets can contain different signature schemes.
+/// Ed25519 is one of the most used ones across blockchains, and hence it's importance to be added.
+/// For further reference, visit: https://ed25519.cr.yp.to
+/// # Cost
+///
+/// `base + write_register_base + write_register_byte * num_bytes + ed25519_verify_base + ed25519_verify_byte * num_bytes`
+pub fn ed25519_verify(
+    &mut self,
+    sig_len: u64,
+    sig_ptr: u64,
+    msg_len: u64,
+    msg_ptr: u64,
+    pub_key_len: u64,
+    pub_key_ptr: u64,
+) -> Result<u64>;
+```
+
+And a `rust-sdk` possible implementation could look like this:
+
+```rs
+
+pub fn ed25519_verify(sig: &ed25519::Signature, msg: &[u8], pub_key: &ed25519::Public) -> bool;
+
+```
+
+Once this NEP is approved and integrated, these functions will be available in the `near_sdk` crate in the
+`env` module.
+
+## Security Implications (Optional)
+
+We have chosen these crates because they are the ones already integrated into the
+[sp_io](https://docs.rs/sp-io/latest/sp_io/)](https://docs.rs/sp-io/latest/sp_io/)
+crate heavily used in the [substrate](https://github.com/paritytech/substrate) repository.
+
+## Unresolved Issues (Optional)
+
+- What parts of the design do you expect to resolve through the NEP process before this gets merged?
+  Both the function signatures and crates are up for discussion.
```

Comment:
I suggest that we use the library that nearcore already uses for ed25519 signatures. This would avoid introducing new dependencies and ensure that NEAR signatures themselves can be 100% verified in smart contracts

Reply:
As far as I can see the proposal is the same as the one being used today
https://github.com/near/nearcore/blob/master/core/crypto/src/signature.rs#L729-L732
https://github.com/near/NEPs/pull/364/files#diff-1d8966e006ab1c29466f5c4d5ce49a965477f992e7d16c1d032d5b0164096459R38